### PR TITLE
feat(react-positioning): add option `shiftToCoverTarget`

### PR DESCRIPTION
**Pull request by YuanboXue-Amber**
## Description

Add positioning option `shiftToCoverTarget` - when there's not enough available space, it allows the positioned element to shift to cover the target element:

https://github.com/user-attachments/assets/9c4a1cb8-9455-48c0-b0eb-d949f3438b34

### CODE REVIEW

File: apps/public-docsite-v9/src/Concepts/Positioning/PositioningShiftToCoverTarget.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import * as React from 'react';
```
Comment (fabricteam) [inline apps/public-docsite-v9/src/Concepts/Positioning/PositioningShiftToCoverTarget.stories.tsx:1]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://aka.ms/visual-changes-review?organization=uifabric&project=fabricpublic&prId=33468&commitId=68845161bf481a928abae9d5bd5f49ba3634db86&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.Badge Mask RTL.chromium.png|9| Changed|
|Avatar Converged.badgeMask.chromium.png|28| Changed|

</details>


  <details><summary><b>Drawer</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Drawer.Full Overlay RTL.chromium.png|1167| Changed|
|Drawer.Full Overlay High Contrast.chromium.png|2244| Changed|
|Drawer.Full Overlay Dark Mode.chromium.png|2798| Changed|

</details>


  <details><summary><b>Positioning</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Positioning.shiftToCoverTarget with autoSize and async content.chromium.png|0| Added|
|Positioning.shiftToCoverTarget with autoSize.chromium.png|0| Added|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-positioning/src/middleware/shift.ts [hunk 2]
```diff
@@ -14,10 +14,22 @@ export interface ShiftMiddlewareOptions
...
+    container,
+    overflowBoundaryPadding,
+    isRtl,
+  } = options;
 
   return baseShift({
     ...(hasScrollableElement && { boundary: 'clippingAncestors' }),
+    ...(shiftToCoverTarget && {
+      crossAxis: true,
+      limiter: limitShift({ crossAxis: true, mainAxis: false }),
```
Comment (ling1726) [inline packages/react-components/react-positioning/src/middleware/shift.ts:31]: would you be able to test what the consequences of applying `shiftToCoverTarget` are to this example? https://react.fluentui.dev/?path=/docs/components-menu-menu--docs#nested-submenus-responsiveness

Comment (ling1726) [reply packages/react-components/react-positioning/src/middleware/shift.ts:31]: And submenus in general actually

File: apps/vr-tests-react-components/src/stories/Positioning/Positioning.stories.tsx [hunk 3]
```diff
@@ -843,6 +843,70 @@ const TargetDisplayNone = () => {
   );
 };
 
+const ShiftToCoverTargetWithAutoSize = () => {
+  const styles = useStyles();
+  const [overflowBoundary, setOverflowBoundary] = React.useState<HTMLDivElement | null>(null);
+  const { containerRef, targetRef } = usePositioning({
+    position: 'below',
```
Comment (ling1726) [inline apps/vr-tests-react-components/src/stories/Positioning/Positioning.stories.tsx:850]: Let's add a test for horizontal axis too

File: apps/public-docsite-v9/src/Concepts/Positioning/PositioningShiftToCoverTarget.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    overflow: 'auto',
+  },
+  trigger: {
+    display: 'block',
+    width: '150px',
+    margin: '200px auto',
+  },
+});
+
+export const CoverTargetForSmallViewport = () => {
```
Comment (ling1726) [inline apps/public-docsite-v9/src/Concepts/Positioning/PositioningShiftToCoverTarget.stories.tsx:28]: Let's add a toggle to switch to an example with horizontal main axis and update the width of a container

File: packages/react-components/react-positioning/src/middleware/shift.ts (hunk 2 above)
Comment (YuanboXue-Amber) [reply packages/react-components/react-positioning/src/middleware/shift.ts:31]: It works as expected, submenu falls back to above when boundary width gets smaller: 
Screenshot 2024-12-16 at 15 13 01

And submenu can shift to cover the trigger item when the height gets smaller:
image

File: apps/public-docsite-v9/src/Concepts/Positioning/PositioningShiftToCoverTarget.stories.tsx (hunk 4 above)
Comment (YuanboXue-Amber) [reply apps/public-docsite-v9/src/Concepts/Positioning/PositioningShiftToCoverTarget.stories.tsx:64]: I added an option to toggle menu position, and made the boundary resizable:
Screenshot 2024-12-16 at 16 01 39

File: apps/vr-tests-react-components/src/stories/Positioning/Positioning.stories.tsx (hunk 3 above)
Comment (YuanboXue-Amber) [reply apps/vr-tests-react-components/src/stories/Positioning/Positioning.stories.tsx:850]: Added test:
Screenshot 2024-12-16 at 16 21 31 
=> 
Screenshot 2024-12-16 at 16 21 35

**Review by ling1726 — APPROVED** [review]

**Pull request merged**